Pass information about resource loads from WebPageProxy through the extensions controller machinery (needed for webRequest support)

b-weinstein opened the pull request:
#### 2afc03180fc116298060b5b592282c471f29225b
<pre>
Pass information about resource loads from WebPageProxy through the extensions controller machinery (needed for webRequest support)
<a href="https://bugs.webkit.org/show_bug.cgi?id=267882">https://bugs.webkit.org/show_bug.cgi?id=267882</a>
<a href="https://rdar.apple.com/114823223">rdar://114823223</a>

Reviewed by Timothy Hatcher.

The events still aren&apos;t fired in the extensions&apos; web process yet, but they will be in the next PR.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::hasPermissionToSendWebRequestEvent): Check that:
- The extension can see the tab
- The extension has access to the page URL
- The extension has access to the resource URL
- The extension has access to the URL specified in the ResourceLoadInfo object
(WebKit::WebExtensionContext::resourceLoadDidSendRequest): Send a message to the WebExtensionContextProxy.
(WebKit::WebExtensionContext::resourceLoadDidPerformHTTPRedirection): Ditto.
(WebKit::WebExtensionContext::resourceLoadDidReceiveChallenge): Ditto.
(WebKit::WebExtensionContext::resourceLoadDidReceiveResponse): Ditto.
(WebKit::WebExtensionContext::resourceLoadDidCompleteWithError): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::resourceLoadDidSendRequest): Call into each WebExtensionContext.
(WebKit::WebExtensionController::resourceLoadDidPerformHTTPRedirection): Ditto.
(WebKit::WebExtensionController::resourceLoadDidReceiveChallenge): Ditto.
(WebKit::WebExtensionController::resourceLoadDidReceiveResponse): Ditto.
(WebKit::WebExtensionController::resourceLoadDidCompleteWithError): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::processes const): Update the method to get all processes for an event to take a set
of events.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::sendToProcessesForEvents):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::resourceLoadDidSendRequest): Call into the WebExtensionController if it exists.
(WebKit::WebPageProxy::resourceLoadDidPerformHTTPRedirection): Ditto.
(WebKit::WebPageProxy::resourceLoadDidReceiveChallenge): Ditto.
(WebKit::WebPageProxy::resourceLoadDidReceiveResponse): Ditto.
(WebKit::WebPageProxy::resourceLoadDidCompleteWithError): Ditto.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
(WebKit::WebExtensionContextProxy::resourceLoadDidSendRequest): Add a FIXME to implement this.
(WebKit::WebExtensionContextProxy::resourceLoadDidPerformHTTPRedirection): Ditto.
(WebKit::WebExtensionContextProxy::resourceLoadDidReceiveChallenge): Ditto.
(WebKit::WebExtensionContextProxy::resourceLoadDidReceiveResponse): Ditto.
(WebKit::WebExtensionContextProxy::resourceLoadDidCompleteWithError): Ditto.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:

Canonical link: <a href="https://commits.webkit.org/273335@main">https://commits.webkit.org/273335@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f1f854a1f09e936a3a684670267dfcdb580cdf0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35092 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14019 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37212 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37852 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/31691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16398 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11089 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35637 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11870 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31316 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10397 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/10462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39099 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31934 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31743 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10572 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8503 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/34459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/12357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8041 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11100 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/11419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1874,6 +1878,93 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 
+// MARK: webRequest
+
+bool WebExtensionContext::hasPermissionToSendWebRequestEvent(WebExtensionTab *tab, const URL& resourceURL, const ResourceLoadInfo& loadInfo)
```
`WebExtensionTab* tab`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1874,6 +1878,93 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+
+    auto tabURL = tab->url();
+    if (tabURL.isValid() && !hasPermission(tabURL, tab))
+        return false;
```
```
if (!tab->extensionHasPermission())
    return false;
```

xeenon replied:
`extensionHasPermission()` also checks for private browsing access.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1874,6 +1878,93 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    if (tabURL.isValid() && !hasPermission(tabURL, tab))
+        return false;
+
+    if (resourceURL.isValid() && !hasPermission(resourceURL, tab))
```
Is `isValid()` needed? `hasPermission()` already does an `isEmpty()` check and returns `false`.

b-weinstein replied:
I missed that it did the isEmpty check. I think we can get rid of the isValid check.

b-weinstein replied:
Actually - is the URL is empty, `hasPermission` will return false. We want to return true in that case, so I want to keep the check.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1874,6 +1878,93 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+        return false;
+
+    const URL& resourceLoadURL = loadInfo.originalURL;
+    if (resourceLoadURL.isValid() && !hasPermission(resourceLoadURL, tab))
```
Ditto for `isValid()`.

b-weinstein replied:
Since an empty URL would lead to `hasPermission` returning false, I think I want to keep this check.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1874,6 +1878,93 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+
+    auto eventTypes = { WebExtensionEventListenerType::WebRequestOnBeforeRequest, WebExtensionEventListenerType::WebRequestOnBeforeSendHeaders, WebExtensionEventListenerType::WebRequestOnSendHeaders };
+    wakeUpBackgroundContentIfNecessaryToFireEvents(eventTypes, [&] {
+        sendToProcessesForEvents(eventTypes, Messages::WebExtensionContextProxy::ResourceLoadDidSendRequest(tab->identifier(), tab->window()->identifier(), request, loadInfo));
```
`tab->window()` can be `null` in rare cases. Will need to check it.

b-weinstein replied:
I'll adopt something like:

`window ? window->identifier() : WebExtensionWindowConstants::NoneIdentifier;`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1874,6 +1878,93 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+
+    auto eventTypes = { WebExtensionEventListenerType::WebRequestOnHeadersReceived, WebExtensionEventListenerType::WebRequestOnBeforeRequest };
+    wakeUpBackgroundContentIfNecessaryToFireEvents(eventTypes, [&] {
+        sendToProcessesForEvents(eventTypes, Messages::WebExtensionContextProxy::ResourceLoadDidSendRequest(tab->identifier(), tab->window()->identifier(), request, loadInfo));
```
Ditto for `tab->window()`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1874,6 +1878,93 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+
+    auto eventTypes = { WebExtensionEventListenerType::WebRequestOnAuthRequired };
+    wakeUpBackgroundContentIfNecessaryToFireEvents(eventTypes, [&] {
+        sendToProcessesForEvents(eventTypes, Messages::WebExtensionContextProxy::ResourceLoadDidReceiveChallenge(tab->identifier(), tab->window()->identifier(), challenge, loadInfo));
```
Ditto for `tab->window()`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1874,6 +1878,93 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+
+    auto eventTypes = { WebExtensionEventListenerType::WebRequestOnHeadersReceived, WebExtensionEventListenerType::WebRequestOnResponseStarted };
+    wakeUpBackgroundContentIfNecessaryToFireEvents(eventTypes, [&] {
+        sendToProcessesForEvents(eventTypes, Messages::WebExtensionContextProxy::ResourceLoadDidReceiveResponse(tab->identifier(), tab->window()->identifier(), response, loadInfo));
```
Ditto for `tab->window()`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1874,6 +1878,93 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+
+    auto eventTypes = { WebExtensionEventListenerType::WebRequestOnErrorOccurred, WebExtensionEventListenerType::WebRequestOnCompleted };
+    wakeUpBackgroundContentIfNecessaryToFireEvents(eventTypes, [&] {
+        sendToProcessesForEvents(eventTypes, Messages::WebExtensionContextProxy::ResourceLoadDidCompleteWithError(tab->identifier(), tab->window()->identifier(), response, error, loadInfo));
```
Ditto for `tab->window()`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -332,6 +332,8 @@
 }
 #endif
 
+// MARK: webNavigation
+
```
I had removed this, since it is used for other things now.

b-weinstein committed:
Pass information about resource loads from WebPageProxy through the extensions controller machinery (needed for webRequest support)
https://bugs.webkit.org/show_bug.cgi?id=267882
rdar://114823223

Reviewed by Timothy Hatcher.

The events still aren't fired in the extensions' web process yet, but they will be in the next PR.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::hasPermissionToSendWebRequestEvent): Check that:
- The extension can see the tab
- The extension has access to the page URL
- The extension has access to the resource URL
- The extension has access to the URL specified in the ResourceLoadInfo object
(WebKit::WebExtensionContext::resourceLoadDidSendRequest): Send a message to the WebExtensionContextProxy.
(WebKit::WebExtensionContext::resourceLoadDidPerformHTTPRedirection): Ditto.
(WebKit::WebExtensionContext::resourceLoadDidReceiveChallenge): Ditto.
(WebKit::WebExtensionContext::resourceLoadDidReceiveResponse): Ditto.
(WebKit::WebExtensionContext::resourceLoadDidCompleteWithError): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::resourceLoadDidSendRequest): Call into each WebExtensionContext.
(WebKit::WebExtensionController::resourceLoadDidPerformHTTPRedirection): Ditto.
(WebKit::WebExtensionController::resourceLoadDidReceiveChallenge): Ditto.
(WebKit::WebExtensionController::resourceLoadDidReceiveResponse): Ditto.
(WebKit::WebExtensionController::resourceLoadDidCompleteWithError): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::processes const): Update the method to get all processes for an event to take a set
of events.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::sendToProcessesForEvents):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::resourceLoadDidSendRequest): Call into the WebExtensionController if it exists.
(WebKit::WebPageProxy::resourceLoadDidPerformHTTPRedirection): Ditto.
(WebKit::WebPageProxy::resourceLoadDidReceiveChallenge): Ditto.
(WebKit::WebPageProxy::resourceLoadDidReceiveResponse): Ditto.
(WebKit::WebPageProxy::resourceLoadDidCompleteWithError): Ditto.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
(WebKit::WebExtensionContextProxy::resourceLoadDidSendRequest): Add a FIXME to implement this.
(WebKit::WebExtensionContextProxy::resourceLoadDidPerformHTTPRedirection): Ditto.
(WebKit::WebExtensionContextProxy::resourceLoadDidReceiveChallenge): Ditto.
(WebKit::WebExtensionContextProxy::resourceLoadDidReceiveResponse): Ditto.
(WebKit::WebExtensionContextProxy::resourceLoadDidCompleteWithError): Ditto.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:

Canonical link: https://commits.webkit.org/273335@main
